Added patreon provider - tested and working

Tetedeiench opened the pull request:
<!--
Thanks for your interest in the project. Bugs filed and PRs submitted are appreciated!

Please make sure that you are familiar with and follow the Code of Conduct for
this project (found in the CODE_OF_CONDUCT.md file).

Also, please make sure you're familiar with and follow the instructions in the
contributing guidelines (found in the CONTRIBUTING.md file).

If you're new to contributing to open source projects, you might find this free
video course helpful: https://kcd.im/pull-request

Please fill out the information below to expedite the review and (hopefully)
merge of your pull request!
-->

<!-- What changes are being made? (What feature/bug is being fixed here?) -->

## Reasoning 💡
As per https://github.com/nextauthjs/next-auth/discussions/3519 , the patreon provider was missing. I added one. Make sure you review thoroughly what I'm doing, despite the changes being simple. I'm updating the docs right after this PR, as they have moved to a separate repo.

<!-- What changes are being made? What feature/bug is being fixed here? -->

## Checklist 🧢
Added Patreon provider

<!-- Feel free cross items ( like this `~[] item~` ) if they're irrelevant to your changes.

To check an item, place an `x` in the box like so: `- [x] Documentation`. -->

- [ x ] Documentation
- [ ] Tests
- [ ] Ready to be merged

<!-- In your opinion, is this ready to be merged as soon as it's reviewed? -->

## Affected issues 🎟
https://github.com/nextauthjs/next-auth/discussions/3519

<!--
Please [scout and link issues](https://github.com/nextauthjs/next-auth/issues) that might be solved by this PR.

If you write `"Fixes"` or `"Closes"` before the issue link like so:

```
Fixes #359
```

the connected issue will be automatically closed once the PR is merged and hence help with maintenance of the library 😊

-->

Tetedeiench committed:
Added patreon provider - tested and working

balazsorban44 commented on app/.env.local.example:
```diff
@@ -1,45 +0,0 @@
-# Rename file to .env.local (or .env) and populate values
```
Why was this file removed?

Tetedeiench replied:
Me being a n00b. I just renamed it, and didn't copy it.

balazsorban44 commented on package-lock.json:
```diff
@@ -5,7 +5,6 @@
   "requires": true,
   "packages": {
     "": {
-      "name": "next-auth",
```
What happened here? Could we revert this change?

Tetedeiench replied:
We can, I just ran npm install, sorry.

balazsorban44 commented on src/providers/patreon.js:
```diff
@@ -0,0 +1,27 @@
+        name: 'Patreon',
+        type: 'oauth',
+        version: '2.0',
+        authorization: { 
+          url : 'https://www.patreon.com/oauth2/authorize?response_type=code',
+          params: { scope: 'identity identity[email] identity.memberships', grant_type: 'authorization_code' }
+        },
```
```suggestion
        authorization: { 
          url : 'https://www.patreon.com/oauth2/authorize',
          params: { scope: 'identity identity[email] identity.memberships' }
        },
```

balazsorban44 commented on src/providers/patreon.js:
```diff
@@ -0,0 +1,27 @@
+        type: 'oauth',
+        version: '2.0',
+        authorization: { 
+          url : 'https://www.patreon.com/oauth2/authorize?response_type=code',
```
```suggestion
          url : 'https://www.patreon.com/oauth2/authorize',
```

I believe this should be unnecessary

balazsorban44 commented on src/providers/patreon.js:
```diff
@@ -0,0 +1,27 @@
+        version: '2.0',
+        authorization: { 
+          url : 'https://www.patreon.com/oauth2/authorize?response_type=code',
+          params: { scope: 'identity identity[email] identity.memberships', grant_type: 'authorization_code' }
```
```suggestion
          params: { scope: 'identity identity[email] identity.memberships' }
```

I believe `grant_type` is set by default.

Also, what's the reason for the `identity.memberships` scope?

Do we need this to get either of these `id`, `name`, `image`, `email`?

If not, it should be best if we set as few scopes as possible by default.

Tetedeiench replied:
Indeed, I'll remove the unnecessary scope

balazsorban44 commented on src/providers/patreon.js:
```diff
@@ -0,0 +1,27 @@
+          url : 'https://www.patreon.com/oauth2/authorize?response_type=code',
+          params: { scope: 'identity identity[email] identity.memberships', grant_type: 'authorization_code' }
+        },
+        token : {
+          url : 'https://www.patreon.com/api/oauth2/token',
+        },
+        userinfo : {
+          url: 'https://www.patreon.com/api/oauth2/api/current_user'
+        },      
```
```suggestion
        token: "https://www.patreon.com/api/oauth2/token",
        userinfo: "https://www.patreon.com/api/oauth2/api/current_user",
```

Just some simplification.

balazsorban44 commented on src/providers/patreon.js:
```diff
@@ -0,0 +1,27 @@
+        userinfo : {
+          url: 'https://www.patreon.com/api/oauth2/api/current_user'
+        },      
+        profile: (profile) => {
+          return {
+            id: profile.data.id,
+            name: profile.data.attributes.full_name,
+            email: profile.data.attributes.email
+          }
+        },
```
```suggestion
        profile(profile) {
          return {
            id: profile.data.id,
            name: profile.data.attributes.full_name,
            email: profile.data.attributes.email,
            image: null,
          }
        },
```

So we don't receive a profile image/picture? Could you check? If so, let's set `image: null`.

balazsorban44 reviewed (changes requested):
Thank you for this PR! I added some comments but generally looks good to me.

I was wondering if you were interested in converting this to TypeScript? It does not require much, see an example here: https://github.com/nextauthjs/next-auth/blob/main/src/providers/auth0.ts

Basically, an added interface to hint the user what they can receive through the `profile` callback.

Could you share a screenshot of sorts that shows a logged-in user as some kind of proof we can refer to later on? :pray:

Thank you once again for your contribution! :green_heart:

Tetedeiench committed:
Update src/providers/patreon.js

Co-authored-by: Balázs Orbán <info@balazsorban.com>

Tetedeiench committed:
Update src/providers/patreon.js

Co-authored-by: Balázs Orbán <info@balazsorban.com>

Tetedeiench committed:
Update src/providers/patreon.js

Co-authored-by: Balázs Orbán <info@balazsorban.com>

Tetedeiench committed:
Update src/providers/patreon.js

Co-authored-by: Balázs Orbán <info@balazsorban.com>

Tetedeiench committed:
Switched to TS, restore .env.local, restore package.json as per comments on the PR

Tetedeiench commented:
Hi,

Thanks for your review ! Sorry for having so many mistakes.

I switched the provider to Typescript. I did my best to restore the files to their original states when necessary.

As for the proof, here's a gif of the full workflow running locally (with the address to show the URLs) using the latest version, switched to typescript : 
![patreon login proof](https://user-images.githubusercontent.com/23140774/148646133-5bb0e9bc-936c-4958-b154-a55ac3f76ec2.gif)

Let me know if there's still things that need some work ! And again, sorry for the poor quality.

balazsorban44 committed:
chore: ran Prettier

balazsorban44 commented:
A stellar contribution, thank you! :green_heart:
